cmd/kubeadm/app/cmd/phases/workflow/runner.go

```diff
@@ -372,6 +372,12 @@ func (e *Runner) BindToCommand(cmd *cobra.Command) {
 		// if this phase has children (not a leaf) it doesn't accept any args
 		if len(p.Phases) > 0 {
 			phaseCmd.Args = cobra.NoArgs
+		} else {
+			if p.Args == nil {
+				phaseCmd.Args = cmd.Args
```

Comment:
Is there any reason to do this? I am not a particular fan of imposing `cobra.MaximumNArgs(1)` in this way.

Reply:
? I'm not sure I understand this. I am saying here, if you did not set any Args in the phase, to just use the parent commands arg checker. right?

Reply:
Also, I believe this is what @fabriziopandini wanted https://github.com/kubernetes/kubeadm/issues/1375#issuecomment-468949883

Reply:
I just took the `join` case, for `init` this would be `cobra.NoArgs`. This means that we need to test all the phases to check for CLI breakages (especially the init phases).

The original proposal by @fabriziopandini (in kubernetes/kubeadm#1375) states:
> More specifically, we would like to have leaf phases without discovery flags using Args: cobra.NoArgs (while the top-level command/all the other leaf phases will use Args: cobra.MaximumNArgs(1))

I agree with this, except for the fact, that `init`s top level cmd should use `cobra.NoArgs` too (which it does ATM).

The way you do it gets the job done, but I distaste the implicit behavior of this. Without full test coverage for all possible phases it can can cause problems in the future.
What I think is a bit more appropriate with respect to the above concern is to make stuff explicit - don't inherit anything from the parent command, but explicitly add the positional args verifier per phase. This will make the things more verbose though and the patch larger in size.

Reply:
I understand @rosti  point that implicit behavior is ... not explicit when you read the code

I considered three options 
a) to apply by default the parent arg validation of parent phase to leaf phases (current implementation)
b) to apply a predefined default to all phases e.g. NoArgs 
c) to not apply any default for leaf phases, but to set args validation only if requested

The option c) goes in the direction of being explicit, but in the end, you have to set args for all phases, and I don't like the idea to add more knobs to all the phases (we already have eeetooomuch flags)

I discarded b) because choosing a default is tricky (why NoArgs instead of something else?)

So I'm back again to a), the current implementation, that IMO makes sense because phases are pieces of the main command,  and so it is very likely that the args value of the parent command will make sense for leaf phases as well. In practices:

- init has noArgs, and all the init phases will get NoArgs.
- join has MaximumNArgs(1), this makes sense for many join phases as well, but you can set NoArgs by exception where this does not make sense: control-plane-prepare/control-plane, control-plane-join/*

That means that with this approach  we are changing only 4 phases over 20/30 phases currently in place

Reply:
The 4 select phases have been updated to `NoArg` I will squash and push when you all give the word. (if you like it that is.)

Reply:
Ok, let's go with inheriting then.
+1 for some of the control-plane join phases to use NoArgs.